src/range-to-string.js

```diff
@@ -0,0 +1,68 @@
+/* global Node */
+
+/**
+ * Return the next node after `node` in a tree order traversal of the document.
+ */
+function nextNode(node, skipChildren) {
+  if (!skipChildren && node.firstChild) {
+    return node.firstChild
+  } else if (node.nextSibling) {
+    return node.nextSibling
+  }
+  while (node.parentNode) {
+    node = node.parentNode
+    if (node.nextSibling) {
+      return node.nextSibling
+    }
+  }
+  /* istanbul ignore next */
+  return null
+}
+
+function firstNode(range) {
+  if (range.startContainer.nodeType === Node.ELEMENT_NODE) {
+    const node = range.startContainer.childNodes[range.startOffset]
+    return node || range.startContainer
```

Comment:
> In practice, I think I've come across many times when browsers generate ranges with an
> offset that is at the end of the startContainer

Ah, I see your point. You're right, this will return the wrong value if the range starts at the end of a non-empty Element. I've made the change suggested and added a test case.

> Actually, thinking about this just now makes me realize that it's possible to create a range with a start that is past its end, and collapsed will be true

When calling `range.setStart` with a point past the current end, the end is moved to the new start position. Likewise, calling `range.setEnd` with a point before the current start will move the start to the new end position.

I have added a test case for this.